### PR TITLE
PLU-283: [INLINE-TILES-6] allow creating of tile columns inline

### PR DESCRIPTION
**Pull request by pregnantboy**
## TL;DR
This PR allows the creation of new columns by directly typing in the dropdown field.

## Why?
Since creating of columns is more common, we dont wanna keep bringing up the modal as it's a bit distracting.

 ## What changed?
- Add new mutation `UpdateTable` to the useCreateNewOption hook. Also updates the hook to accept the current form field parameters.
- Modified `SingleSelectProvider` to add option to create the new column. This is similar to the `freeSolo` option except that it does not create the option immediately. Instead, it calls the `addNew?.onSelected()` callback.
- Fix minor styling issue with DropdownItem to vertically center the icon

## How to test?
- Optional: throttle network speed
- Using the Tile > Create row action, trying typing a new column name and select the `Create new column` option.
- You should see that it will show `Creating...` with a spinner and the new column is created and selected.

### CODE REVIEW

**Comment by pregnantboy** [comment]
* **#663** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/663?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#661** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/661?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#660** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/660?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#659** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/659?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#657** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/657?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#648** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/648?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#647** <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/647?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `develop-v2`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @pregnantboy and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/frontend/src/components/SingleSelect/SingleSelectProvider.tsx [hunk 1]
```diff
@@ -171,21 +174,45 @@ export const SingleSelectProvider = ({
...
       }
     },
     [getItemByValue],
   )
 
+  const addInlineNewOption = useCallback(
+    (filteredItems: ComboboxItem<string>[], inputValue?: string) => {
+      if (inputValue?.trim() && !getItemByValue(inputValue)) {
+        filteredItems.push({
+          value: '__ADD_NEW_PLACEHOLDER_VALUE__', // this is not referenced anywhere else
```
Comment (ogp-weeloong) [inline packages/frontend/src/components/SingleSelect/SingleSelectProvider.tsx]: nit: since using in more than 1 place, can make it a constant

File: packages/frontend/src/components/ControlledAutocomplete/AddNewOptionModal.tsx [hunk 2]
```diff
@@ -49,6 +54,25 @@ export function useCreateNewOption(setValue: (newValue: string) => void) {
             newValue = data?.createTable.id
             break
           }
+          case 'tiles-createTileRow-columnId': {
+            const tableId = parameters?.tableId
+            if (!tableId) {
+              return
+            }
```
Comment (ogp-weeloong) [inline packages/frontend/src/components/ControlledAutocomplete/AddNewOptionModal.tsx]: can do this to fix type issues

```suggestion
            const tableId = parameters?.tableId as string // confirm plus chop
            if (!tableId) {
              return
            }
```

or alternatively this but tbh this is no better than a hard cast because we're returning silently.

```suggestion
            const tableId = parameters?.tableId
            if (typeof tableId !== 'string' || !tableId) {
              return
            }
```

File: packages/frontend/src/components/ControlledAutocomplete/AddNewOptionModal.tsx [hunk 3]
```diff
@@ -49,6 +54,25 @@ export function useCreateNewOption(setValue: (newValue: string) => void) {
...
+                  id: tableId,
+                  addedColumns: [inputValue],
+                },
+              },
+            })
+            const newColumns = data?.updateTable?.columns
+            newValue = newColumns.find(
+              (column: ITableColumnMetadata) => column.name === inputValue,
+            )?.id
+            break
```
Comment (ogp-weeloong) [inline packages/frontend/src/components/ControlledAutocomplete/AddNewOptionModal.tsx]:
```suggestion
            const newColumns = data?.updateTable?.columns ?? []
            newValue = newColumns.find(
              (column) => column.name === inputValue,
            )?.id
            break
```

File: packages/frontend/src/components/ControlledAutocomplete/index.tsx [hunk 4]
```diff
@@ -149,7 +164,10 @@ function ControlledAutocomplete(
                 ? {
```
Comment (ogp-weeloong) [inline packages/frontend/src/components/ControlledAutocomplete/index.tsx:165]: <i class='graphite__hidden'>[Re: line 152]<br/><br/></i>

can we also add 
```
isReadOnly={isCreatingNewOption}
```<!--__GRAPHITE_HTML_TAG_START__--><span class='graphite__hidden'><br/><br/>See this comment inline on <a href="https://app.graphite.dev/github/pr/opengovsg/plumber/661?utm_source=unchanged-line-comment">Graphite</a>.</span>

<!--__GRAPHITE_HTML_TAG_END__-->

**Review by ogp-weeloong — APPROVED** [review]
lgtm. tested:
- can create new inline column
- compatible with allowArbitrary (although i dunno when we'll ever use them together)

File: packages/frontend/src/components/ControlledAutocomplete/AddNewOptionModal.tsx (hunk 3 above)
Comment (pregnantboy) [reply packages/frontend/src/components/ControlledAutocomplete/AddNewOptionModal.tsx]: fixed

File: packages/frontend/src/components/ControlledAutocomplete/AddNewOptionModal.tsx (hunk 2 above)
Comment (pregnantboy) [reply packages/frontend/src/components/ControlledAutocomplete/AddNewOptionModal.tsx]: fixed

**Comment by pregnantboy** [comment]
### Merge activity

* **Jul 4, 3:34 AM EDT**: @pregnantboy started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/opengovsg/plumber/661).
* **Jul 4, 3:43 AM EDT**: [Graphite](https://app.graphite.dev/github/pr/opengovsg/plumber/661) rebased this pull request as part of a merge.
* **Jul 4, 3:44 AM EDT**: @pregnantboy merged this pull request with [Graphite](https://app.graphite.dev/github/pr/opengovsg/plumber/661).
